src/plugins/main.js

```diff
@@ -118,25 +120,44 @@ export default async (context) => {
     }
   }
 
-  const loadAndSetLocale = async locale => {
-    const { loadLanguageAsync } = require('./utils')
+  const loadAndSetLocale = async (newLocale, init = false) => {
+    // Abort if different domains option enabled
+    if (app.i18n.differentDomains) {
+      return
+    }
 
-    let messages
+    // Abort if newLocale did not change
+    if (newLocale === app.i18n.locale) {
+      return
+    }
 
+    const oldLocale = app.i18n.locale
+    app.i18n.beforeLanguageSwitch(oldLocale, newLocale)
+    if (useCookie && !init) {
+      app.i18n.setLocaleCookie(newLocale)
+    }
+    // Lazy-loading enabled
     if (lazy) {
-      // Load fallback locale.
-      if (app.i18n.fallbackLocale && locale !== app.i18n.fallbackLocale) {
-        await loadLanguageAsync(context, app.i18n.fallbackLocale)
-      }
-
-      messages = await loadLanguageAsync(context, locale)
+      const { loadLanguageAsync } = require('./utils')
+      const messages = await loadLanguageAsync(context, newLocale)
+      app.i18n.locale = newLocale
+      app.i18n.onLanguageSwitched(oldLocale, newLocale)
+      await syncVuex(newLocale, messages)
     } else {
-      messages = app.i18n.getLocaleMessage(locale)
+      // Lazy-loading disabled
+      app.i18n.locale = newLocale
+      app.i18n.onLanguageSwitched(oldLocale, newLocale)
+      await syncVuex(newLocale, app.i18n.getLocaleMessage(newLocale))
     }
 
-    app.i18n.locale = locale
-    await syncVuex(locale, messages)
+    const routeName = route && route.name ? app.getRouteBaseName(route) : 'index'
+
+    if (!init && STRATEGY !== STRATEGIES.NO_PREFIX) {
+      redirect(app.localePath(Object.assign({}, route , {
```

Comment:
(Just some thoughts that don't have to be handled now.)

On the client side we have an option of using VueRouter to push/replace the route. That would make more sense if that API is called manually I guess. But maybe not when middleware calls it (even on the client)... Something to investigate.

Reply:
If I'm not wrong, it was already behaving with redirect prior to this PR. But I agree with you, it may be a good idea to investigate further after merging.

Reply:
"Behaving" yes but I'm thinking of behaving better since that API is now public so it will be called from user action and it would be better to use `router.replace` rather than `location.href` which that API does on the client side.